### PR TITLE
Introduce GitHub markdown editor, keep EasyMDE as fallback

### LINKED ISSUE #10729: Replace EasyMDE with textarea
SimpleMDE is unmaintained and in my opionion very bloated for what it does. It ships with a full version of CodeMirror (a code editor) embedded which enables markdown syntax highlight in the editor, but I think its a neglible feature that is mostly unnecessary as comments are meant to be viewed in rendered mode primarily.

I suggest we switch to a plain `<textarea>` like the ones Github and Gogs use. Some buttons to insert pre-generated snippets at the cursor can be provided. Emoji can be supported via `:token:` to Unicode  replacement or directly as Unicode as always. Image-based custom emoji can not render in it, but they may in the rendered view (thought Unicode should be preferred).

### PR DESCRIPTION
**Pull request by wxiaoguang**
The first step of the plan

* #23290

Thanks to @silverwind for the first try in #15394 . Close #10729 and a lot of related issues.

The EasyMDE is not removed, now it works as a fallback, users can switch between these two editors.

Editor list:

* Issue / PR comment
* Issue / PR comment edit
* Issue / PR comment quote reply
* PR diff view, inline comment
* PR diff view, inline comment edit
* PR diff view, inline comment quote reply
* Release editor
* Wiki editor

Some editors have attached dropzone

Screenshots:

<details>

image

image

image

image

</details>

### CODE REVIEW

**Comment by silverwind** [comment]
> The EasyMDE is not removed, now it works as a fallback, users can switch between these two editors.

I'd say we should remove it eventually, maybe not initially but I see no benefit in keeping this buggy editor around any longer.

File: web_src/css/editor-markdown.css [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+  user-select: none;
+  padding: 5px;
+  cursor: pointer;
+}
+
+.combo-markdown-editor .markdown-text-editor{
+  display: block;
+  width: 100%;
+  height: 200px;
+}
```
Comment (silverwind) [inline web_src/css/editor-markdown.css:22]: Bit of formatting please, empty line between rules, space before `{`.

File: templates/devtest/gitea-ui.tmpl [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+	</div>
+
+	<div>
+		<span data-tooltip-content="test tooltip">text with tooltip</span>
+	</div>
+
+	{{template "shared/combomarkdowneditor" .}}
+
+</div>
+
```
Comment (silverwind) [inline templates/devtest/gitea-ui.tmpl]:
```suggestion
<div class="page-content devtest">
	<div>
		<gitea-origin-url data-url="test/url"></gitea-origin-url>
		<gitea-origin-url data-url="/test/url"></gitea-origin-url>
	</div>
	<div>
		<span data-tooltip-content="test tooltip">text with tooltip</span>
	</div>
	{{template "shared/combomarkdowneditor" .}}
</div>
```

File: templates/repo/diff/comment_form.tmpl [hunk 3]
```diff
@@ -9,18 +9,16 @@
...
-		</div>
+
+		{{template "shared/combomarkdowneditor" (dict
+			"locale" $.root.locale
+			"MarkdownPreviewUrl" (print $.root.Repository.Link "/markup")
+			"MarkdownPreviewContext" $.root.RepoLink
+			"TextareaName" "content"
+			"TextareaPlaceholder" ($.locale.Tr "repo.diff.comment.placeholder")
+			"DropzoneParentContainer" "form"
+		)}}
```
Comment (silverwind) [inline templates/repo/diff/comment_form.tmpl:20]: This is brilliant, I need to copy that style for future PRs 😉

File: web_src/css/editor-markdown.css (hunk 1 above)
Comment (silverwind) [inline web_src/css/editor-markdown.css]:
```suggestion
.combo-markdown-editor {
  width: 100%;
}

.combo-markdown-editor markdown-toolbar {
  cursor: default;
  display: block;
  padding-bottom: 10px;
}

.combo-markdown-editor .markdown-toolbar-group {
  display: inline-block;
}

.combo-markdown-editor .markdown-toolbar-button {
  user-select: none;
  padding: 5px;
  cursor: pointer;
}

.combo-markdown-editor .markdown-text-editor {
  display: block;
  width: 100%;
  height: 200px;
}
```

**Review by silverwind — APPROVED** [review]
Works well enough, refinement and EasyMDE removal can be done later. I do have a few changes in mind, but let's not delay getting this in.

File: templates/shared/combomarkdowneditor.tmpl [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+			</div>
+		</markdown-toolbar>
+		<textarea class="markdown-text-editor js-quick-submit" name="{{.TextareaName}}" placeholder="{{.TextareaPlaceholder}}">{{.TextareaContent}}</textarea>
+	</div>
+
+	<div class="ui tab markup" data-tab="markdown-previewer">
+		{{.locale.Tr "loading"}}
+	</div>
+
+</div>
```
Comment (silverwind) [inline templates/shared/combomarkdowneditor.tmpl]:
```suggestion
<div {{if .ContainerId}}id="{{.ContainerId}}"{{end}} class="combo-markdown-editor {{.ContainerClasses}}" data-dropzone-parent-container="{{.DropzoneParentContainer}}">
	{{if .MarkdownPreviewUrl}}
	<div class="ui top tabular menu">
		<a class="active item" data-tab="markdown-writer">{{.locale.Tr "write"}}</a>
		<a class="item" data-tab="markdown-previewer" data-preview-url="{{.MarkdownPreviewUrl}}" data-preview-context="{{.MarkdownPreviewContext}}">{{.locale.Tr "preview"}}</a>
	</div>
	{{end}}
	<div class="ui tab active" data-tab="markdown-writer">
		<markdown-toolbar class="gt-df">
			<div class="markdown-toolbar-group">
				<md-header class="markdown-toolbar-button">{{svg "octicon-heading"}}</md-header>
				<md-bold class="markdown-toolbar-button">{{svg "octicon-bold"}}</md-bold>
				<md-italic class="markdown-toolbar-button">{{svg "octicon-italic"}}</md-italic>
			</div>
			<div class="markdown-toolbar-group">
				<md-quote class="markdown-toolbar-button">{{svg "octicon-quote"}}</md-quote>
				<md-code class="markdown-toolbar-button">{{svg "octicon-code"}}</md-code>
				<md-link class="markdown-toolbar-button">{{svg "octicon-link"}}</md-link>
			</div>
			<div class="markdown-toolbar-group">
				<md-unordered-list class="markdown-toolbar-button">{{svg "octicon-list-unordered"}}</md-unordered-list>
				<md-ordered-list class="markdown-toolbar-button">{{svg "octicon-list-ordered"}}</md-ordered-list>
				<md-task-list class="markdown-toolbar-button">{{svg "octicon-tasklist"}}</md-task-list>
			</div>
			<div class="markdown-toolbar-group">
				<md-mention class="markdown-toolbar-button">{{svg "octicon-mention"}}</md-mention>
				<md-ref class="markdown-toolbar-button">{{svg "octicon-cross-reference"}}</md-ref>
			</div>
			<div class="markdown-toolbar-group gt-f1"></div>
			<div class="markdown-toolbar-group">
				<span class="markdown-toolbar-button markdown-switch-easymde">{{svg "octicon-arrow-switch"}}</span>
			</div>
		</markdown-toolbar>
		<textarea class="markdown-text-editor js-quick-submit" name="{{.TextareaName}}" placeholder="{{.TextareaPlaceholder}}">{{.TextareaContent}}</textarea>
	</div>
	<div class="ui tab markup" data-tab="markdown-previewer">
		{{.locale.Tr "loading"}}
	</div>
</div>
```

**Comment by wxiaoguang** [comment]
> > The EasyMDE is not removed, now it works as a fallback, users can switch between these two editors.
> 
> I'd say we should remove it, maybe not initially but I see no benefit in keeping this buggy editor around any longer, so I'd nuke it with this PR.

I also agree that the EasyMDE should be removed in the end, while I think maybe we can leave more time for it.

My opinions are: maybe there are some users who really like the EasyMDE (not everyone user is affected by its bug). If we remove it quickly, these users will be hurt.

Now, the EasyMDE is only the second choice, and there is almost no technical debit of it, I guess this PR's approach could satisfy most users. To remove the EasyMDE feature gracefully, maybe we could do these steps:

1. Show textarea by default, but if a user swiches to EasyMDE, then remember the choice, next time show EasyMDE first.
    * It could be my next PR (if some users need it), to "remember" the referred editor (by LocalStorage)
2. Always show textarea by default, no remember, hide the switch button
    * Only use textarea by default, but if any user really needs the EasyMDE, provide a hidden config option to get it back.
3. Remove the EasyMDE completely if no user complains

**Comment by silverwind** [comment]
Once this is in, we should go through open issues. I estimate we can probably close 30+ issues with this.

**Review by lunny — APPROVED** [review]

**Comment by silverwind** [comment]
One visual change that I would do is move the toolbar to right side of the tabs like I did in my original PR. Maybe put them on separate line when screen space becomes too small.

**Comment by wxiaoguang** [comment]
Conflicts are resolved, and fix some bugs during my test.

**Pull request merged**